### PR TITLE
Make Phone & Email of Contact optional. Simplify some functions in ParserUtil

### PR DESCRIPTION
**Pull request by UncleGrandpa925**

### CODE REVIEW

**Comment by UncleGrandpa925** [comment]
Hi @simonteozw . In this PR I have made the following changes:
- made Email & Phone to be optional (this took quite a lot of time but I have done some manual testing and it seems to be working well) 
- replaced the Task.isValidString by a simple call to String.isBlank (because I think your intention is to check if a string is blank or not)
Please help me review this PR! Thanks a lot (x100)!

File: src/main/java/trackitnus/storage/JsonAdaptedContact.java [hunk 1]
```diff
@@ -74,21 +74,25 @@ public Contact toModelType() throws IllegalValueException {
...
 
+        final Phone modelPhone;
         if (phone == null) {
-            throw new IllegalValueException(String.format(MISSING_FIELD_MESSAGE_FORMAT, Phone.class.getSimpleName()));
+            modelPhone = null;
+        } else {
+            if (!Phone.isValidPhone(phone)) {
+                throw new IllegalValueException(Phone.MESSAGE_CONSTRAINTS);
+            }
+            modelPhone = new Phone(phone);
```
Comment (simonteozw) [inline src/main/java/trackitnus/storage/JsonAdaptedContact.java]: Can we shorten this to 1 `else if {` instead of `else { if {` ?

File: src/main/java/trackitnus/storage/JsonAdaptedContact.java [hunk 2]
```diff
@@ -74,21 +74,25 @@ public Contact toModelType() throws IllegalValueException {
...
-            throw new IllegalValueException(String.format(MISSING_FIELD_MESSAGE_FORMAT, Email.class.getSimpleName()));
-        }
-        if (!Email.isValidEmail(email)) {
-            throw new IllegalValueException(Email.MESSAGE_CONSTRAINTS);
+            modelEmail = null;
+        } else {
+            if (!Email.isValidEmail(email)) {
+                throw new IllegalValueException(Email.MESSAGE_CONSTRAINTS);
+            }
+            modelEmail = new Email(email);
```
Comment (simonteozw) [inline src/main/java/trackitnus/storage/JsonAdaptedContact.java]: Same as phone

File: src/test/java/trackitnus/logic/parser/ParserUtilTest.java [hunk 3]
```diff
@@ -81,25 +82,25 @@ public void parseName_validValueWithWhitespace_returnsTrimmedName() throws Excep
 
     @Test
     public void parsePhone_null_throwsNullPointerException() {
-        Assert.assertThrows(NullPointerException.class, () -> ParserUtil.parsePhone(null));
+        Assert.assertThrows(NullPointerException.class, () -> ParserUtil.parseOptionalPhone(null));
```
Comment (simonteozw) [inline src/test/java/trackitnus/logic/parser/ParserUtilTest.java:85]: Should this be Optional.empty()? Or maybe we should add another test case for that

File: src/test/java/trackitnus/logic/parser/contact/AddContactCommandParserTest.java [hunk 4]
```diff
@@ -90,17 +90,17 @@ public void parse_compulsoryFieldMissing_failure() {
...
+//        assertParseFailure(parser,
+//            ContactCommandTestUtil.NAME_DESC_BOB + ContactCommandTestUtil.VALID_PHONE_BOB
+//                + ContactCommandTestUtil.EMAIL_DESC_BOB + ContactCommandTestUtil.ADDRESS_DESC_BOB,
+//            expectedMessage);
+
+//        // missing email prefix
+//        assertParseFailure(parser,
+//            ContactCommandTestUtil.NAME_DESC_BOB + ContactCommandTestUtil.PHONE_DESC_BOB
+//                + ContactCommandTestUtil.VALID_EMAIL_BOB + ContactCommandTestUtil.ADDRESS_DESC_BOB,
+//            expectedMessage);
```
Comment (simonteozw) [inline src/test/java/trackitnus/logic/parser/contact/AddContactCommandParserTest.java:103]: Why are these lines commented out?

**Review by simonteozw — COMMENTED** [review]
Looks good! Just a few comments

File: src/test/java/trackitnus/logic/parser/contact/AddContactCommandParserTest.java (hunk 4 above)
Comment (UncleGrandpa925) [reply src/test/java/trackitnus/logic/parser/contact/AddContactCommandParserTest.java:103]: because now missing phone is not an error any more, later on I will convert them to positive testcases

File: src/main/java/trackitnus/storage/JsonAdaptedContact.java (hunk 1 above)
Comment (UncleGrandpa925) [reply src/main/java/trackitnus/storage/JsonAdaptedContact.java]: Yeah I think we can!

File: src/test/java/trackitnus/logic/parser/ParserUtilTest.java (hunk 3 above)
Comment (UncleGrandpa925) [reply src/test/java/trackitnus/logic/parser/ParserUtilTest.java:85]: Optional.empty here is actually still valid! null will throw exception!

File: src/main/java/trackitnus/logic/commands/contact/EditContactCommand.java [hunk 5]
```diff
@@ -71,8 +71,8 @@ private static Contact createEditedContact(Contact contactToEdit, EditContactDes
         assert contactToEdit != null;
 
         Name updatedName = editContactDescriptor.getName().orElse(contactToEdit.getName());
-        Phone updatedPhone = editContactDescriptor.getPhone().orElse(contactToEdit.getPhone());
-        Email updatedEmail = editContactDescriptor.getEmail().orElse(contactToEdit.getEmail());
+        Phone updatedPhone = editContactDescriptor.getPhone().orElse(contactToEdit.getPhone().orElse(null));
+        Email updatedEmail = editContactDescriptor.getEmail().orElse(contactToEdit.getEmail().orElse(null));
```
Comment (simonteozw) [inline src/main/java/trackitnus/logic/commands/contact/EditContactCommand.java]: What if the contact originally has a phone number and the user wants to edit to remove it. This will ensure that if the user types `C edit 1 p/` the phone number will remain unchanged when it should be changed to null

File: src/test/java/trackitnus/storage/JsonAdaptedContactTest.java [hunk 6]
```diff
@@ -73,12 +73,12 @@ public void toModelType_invalidEmail_throwsIllegalValueException() {
...
-        JsonAdaptedContact contact = new JsonAdaptedContact(VALID_NAME, VALID_PHONE, null, VALID_TAGS);
-        String expectedMessage = String.format(MISSING_FIELD_MESSAGE_FORMAT, Email.class.getSimpleName());
-        Assert.assertThrows(IllegalValueException.class, expectedMessage, contact::toModelType);
-    }
+//    @Test TODO: modify this to be a positive test
+//    public void toModelType_nullEmail_throwsIllegalValueException() {
+//        JsonAdaptedContact contact = new JsonAdaptedContact(VALID_NAME, VALID_PHONE, null, VALID_TAGS);
+//        String expectedMessage = String.format(MISSING_FIELD_MESSAGE_FORMAT, Email.class.getSimpleName());
+//        Assert.assertThrows(IllegalValueException.class, expectedMessage, contact::toModelType);
+//    }
```
Comment (simonteozw) [inline src/test/java/trackitnus/storage/JsonAdaptedContactTest.java:81]: I think we should create a github issue abt this so we dont forget about it. And include all the lines we commented out so we know what to replace.

**Review by simonteozw — COMMENTED** [review]
Looks good! Just 2 comments. For the EditContactCommand you can refer to my EditTaskCommand.

File: src/test/java/trackitnus/storage/JsonAdaptedContactTest.java (hunk 6 above)
Comment (UncleGrandpa925) [reply src/test/java/trackitnus/storage/JsonAdaptedContactTest.java:81]: Actually in the upcoming PR there will be several more TODOs, it's a common & quick way to note that there are things todo haha since IntelliJ can highlight all todos in the project when you want to!

File: src/main/java/trackitnus/logic/commands/contact/EditContactCommand.java (hunk 5 above)
Comment (UncleGrandpa925) [reply src/main/java/trackitnus/logic/commands/contact/EditContactCommand.java]: Fixed! For the commit I just pushed to fix this I have manually tested deleting phones & emails and it all worked well!

**Review by simonteozw — APPROVED** [review]
LGTM
